python/lsst/pipe/tasks/matchBackgrounds.py

```diff
@@ -19,683 +19,627 @@
-            refInd = self.selectRefExposure(
-                expRefList=expRefList,
-                imageScalerList=imageScalerList,
-                expDatasetType=expDatasetType,
-            )
-            refExpDataRef = expRefList[refInd]
-            refImageScaler = imageScalerList[refInd]
+        # TODO: matching currently done in fluence, not surface brightness,
+        # units.  Current solution for filling in non-overlapping regions
+        # between ref and other images is fitting a plane, but this fails as
+        # plane is typically buried beneath warping Jacobian pattern.  Unsure
+        # how to resolve this issue, but it's critical to resolve for BG-
+        # matching to be viable.
+        if (numExp := len(warps)) < 1:
+            raise TaskError("No exposures to match")
+
+        if self.config.refWarpVisit is None:
+            # Build FFP BG models of each visit
+            visitTractBgs = self._makeTractBackgrounds(warps, skyMap)
+            # Choose a reference visit using those
+            refVisId = self._defineWarps(visitTractBgs)
+        else:
+            self.log.info("Using user-supplied reference visit %d", self.config.refWarpVisit)
+            refVisId = self.config.refWarpVisit
 
-        # refIndSet is the index of all exposures in expDataList that match the reference.
-        # It is used to avoid background-matching an exposure to itself. It is a list
-        # because it is possible (though unlikely) that expDataList will contain duplicates.
-        expKeyList = refExpDataRef.butlerSubset.butler.getKeys(expDatasetType)
-        refMatcher = DataRefMatcher(refExpDataRef.butlerSubset.butler, expDatasetType)
-        refIndSet = set(refMatcher.matchList(ref0=refExpDataRef, refList=expRefList))
+        self.log.info("Matching %d Exposures", numExp)
 
-        if refInd is not None and refInd not in refIndSet:
-            raise RuntimeError("Internal error: selected reference %s not found in expRefList")
+        backgroundInfoList, matchedImageList = self.matchBackgrounds(warps, skyMap, refVisId)
 
-        refExposure = refExpDataRef.get()
-        if refImageScaler is not None:
-            refMI = refExposure.getMaskedImage()
-            refImageScaler.scaleMaskedImage(refMI)
+        # TODO: costly, but consider a post-hoc check on the match quality?
 
-        debugIdKeyList = tuple(set(expKeyList) - set(['tract', 'patch']))
+        return Struct(backgroundInfoList=backgroundInfoList, matchedImageList=matchedImageList)
 
-        self.log.info("Matching %d Exposures", numExp)
+    @timeMethod
+    def _makeTractBackgrounds(self, warps, skyMap, refVisitId=None):
+        """If no reference visit ID is supplied, create full tract models of
+        the backgrounds of all visits.
+        If a reference visit ID is supplied, create full tract models of the
+        difference image backgrounds between all visits and the reference
+        visit.
```

Comment:
This returning a difference background in one case and a non-difference background in another case suggests that some refactoring is necessary.